gradle/documentation/markdown.gradle

```diff
@@ -0,0 +1,172 @@
+  }.join('\n')
+}
+
+configure(subprojects.findAll { it.path == ':lucene' || it.path == ':solr' }) {
+  task markdownToHtml(type: Copy) {
+    filteringCharset = 'UTF-8'
+    includeEmptyDirs = false
+    into project.docroot
+    rename(/\.md$/, '.html')
+    filter(MarkdownFilter)
+  }
+}
+
+configure(project(':lucene')) {
+  markdownToHtml {
+    from('.') {
+      include 'MIGRATE.md'
+      include 'JRE_VERSION_MIGRATION.md'
+      include 'SYSTEM_REQUIREMENTS.md'
+    }
+  }
+  
+  task createDocumentationIndex(type: MarkdownTemplateTask) {
+    def defaultCodecFile = project(':lucene:core').file('src/java/org/apache/lucene/codecs/Codec.java')
+    inputs.file(defaultCodecFile)
+
+    outputFile = file("${project.docroot}/index.html")
+    templateFile = file('site/xsl/index.template.md')
+    
+    binding = {
+      // static Codec defaultCodec   =   LOADER    .   lookup    (   "LuceneXXX"  )   ;
+      def regex = ~/\bdefaultCodec\s*=\s*LOADER\s*\.\s*lookup\s*\(\s*"([^"]+)"\s*\)\s*;/
+      def matcher = regex.matcher(defaultCodecFile.getText('UTF-8'))
+      if (!matcher.find()) {
+        throw GradleException("Cannot determine default codec from file ${defaultCodecFile}")
+      }
+      def majorVersion = project.version.split(/\./)[0] as int;
+      return [
+        defaultCodecPackage : matcher.group(1).toLowerCase(Locale.ROOT),
+        version : project.version,
+        majorVersion : majorVersion,
+        projectList : getListOfProjectsAsMarkdown(':lucene:')
+      ]
+    }
+  }
+}
+
+configure(project(':solr')) {
+  markdownToHtml {
+    from('site') {
+      include '**/*.md'
+    }
+  }
+  
+  task createDocumentationIndex {
+    // nocommit: this needs to be implemented next
+  }
+}
+
+// filter that can be used with the "copy" task of Gradle that transforms Markdown files
+// from source location to HTML (adding HTML header, styling,...)
+class MarkdownFilter extends FilterReader {
+
+  public MarkdownFilter(Reader reader) throws IOException {
+    // this is not really a filter: it reads whole file in ctor,
+    // converts it and provides result downstream as a StringReader
+    super(new StringReader(convert(reader.text)));
+  }
+  
+  public static String convert(String markdownSource) {
+    // first replace LUCENE and SOLR issue numbers with a markdown link
+    markdownSource = markdownSource.replaceAll(/(?s)\b(LUCENE|SOLR)\-\d+\b/,
+      '[$0](https://issues.apache.org/jira/browse/$0)');
+  
+    // convert the markdown
+    MutableDataSet options = new MutableDataSet();
+    options.setFrom(ParserEmulationProfile.MARKDOWN);
+    options.set(Parser.EXTENSIONS, [ AbbreviationExtension.create(), AutolinkExtension.create(), AttributesExtension.create() ]);
+    options.set(HtmlRenderer.RENDER_HEADER_ID, true);
+    options.set(HtmlRenderer.MAX_TRAILING_BLANK_LINES, 0);
+    Document parsed = Parser.builder(options).build().parse(markdownSource);
+
+    StringBuilder html = new StringBuilder('<html>\n<head>\n');
```

Comment:
Would it be better inserting `<!DOCTYPE html>` at the beginning (as a valid html5)?

Reply:
Yes, if we add CSS that's a good idea. With the current unstyled stuff this was the simplest we can do.

If we change that here, we should also do this in tools/ where Ant's script is located.

Reply:
Applying CSS was in my mind. I will add the tag later when I add the style file.

Reply:
I would add CSS in a separate issue. I just want to get everything running first. What do you think about current state?

Reply:
> I would add CSS in a separate issue. 

Yes I think so too. I will open separate issue for that.